### PR TITLE
RBAC Testing an admin role

### PR DESCRIPTION
**Pull request by itzikb-redhat**
First test for RBAC.
Creating a role with creating,listing and getting flightctl resources and a a role binding to a user with no prior permissions.

Making sure the user can:
Create, update, list, get a specific and delete a:
 - Device
 - Fleet
 - Repository

Create and list:
- Enrollment requests
- Events

## Summary by CodeRabbit

- Tests
  - Added comprehensive end-to-end RBAC suites exercising namespace- and cluster-scoped permissions with per-test VM/agent isolation, setup/teardown, and gating to skip when ACM isn’t available.
- Refactor
  - Centralized ACM presence detection into a shared test utility for consistent test gating across suites.
- Chores
  - Expanded test harness and tooling: RBAC and resource lifecycle helpers, YAML label tooling, Kubernetes context utilities, non-admin login helper, repository fetch helper, and an example Repository test manifest.

### CODE REVIEW

File: test/e2e/rbac/rbac_test.go [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+
+type flightCtlResource struct {
+	ResourceType string
+	ResourceName string
+}
+
+var _ = Describe("RBAC Authorization Tests", Label("rbac", "authorization"), func() {
+	var harness *e2e.Harness
+
+	flightCtlNs := os.Getenv("FLIGHTCTL_NS")
```
Comment (siserafin) [inline test/e2e/rbac/rbac_test.go]: Should we add a default value or fail the test if the env var is empty?

File: test/e2e/rbac/rbac_test.go [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+	BeforeEach(func() {
+		harness = e2e.NewTestHarness(suiteCtx)
+		flightCtlResources = []flightCtlResource{}
+	})
+
+	AfterEach(func() {
+		ChangeContext("default")
+		login.LoginToAPIWithToken(harness)
+
+		for _, resource := range flightCtlResources {
```
Comment (siserafin) [inline test/e2e/rbac/rbac_test.go]: Maybe you could use:

```
err := harness.CleanUpAllResources()

File: test/e2e/rbac/rbac_test.go (hunk 1 above)
Comment (itzikb-redhat) [reply test/e2e/rbac/rbac_test.go]: I think it should fail, wdyt?

File: test/e2e/rbac/rbac_test.go [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+			output, err := harness.CleanUpResource(resource.ResourceType, resource.ResourceName)
+			if err != nil {
+				logrus.Errorf("Failed to clean up resource %s of resource type %s: %v", resource.ResourceName, resource.ResourceType, err)
+			} else {
+				logrus.Infof("Cleaned up resource %s of resource type %s: %s", resource.ResourceName, resource.ResourceType, output)
+			}
+		}
+
+		for _, role := range roles {
```
Comment (siserafin) [inline test/e2e/rbac/rbac_test.go]: We could create a local function to cleanup the roles

File: test/e2e/rbac/rbac_test.go (hunk 2 above)
Comment (itzikb-redhat) [reply test/e2e/rbac/rbac_test.go]: Won't that be a problem when running tests in parallel?

File: test/e2e/rbac/rbac_test.go (hunk 3 above)
Comment (itzikb-redhat) [reply test/e2e/rbac/rbac_test.go]: Will do

File: test/e2e/rbac/rbac_test.go (hunk 1 above)
Comment (itzikb-redhat) [reply test/e2e/rbac/rbac_test.go]: I've made it skip

File: test/e2e/rbac/rbac_test.go [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"github.com/sirupsen/logrus"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/yaml"
+)
+
+const (
```
Comment (siserafin) [inline test/e2e/rbac/rbac_test.go:14]: I would set these as constants:
```
const (
	nonAdminUser         = "demouser1"
	adminRoleName        = "rbac-test-admin-role"
	adminRoleBindingName = "rbac-test-admin-role-binding"
)

File: test/e2e/rbac/rbac_test.go [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+
+			adminRoleBinding, err := createRoleBinding(harness.Cluster, flightCtlNs, adminRoleBinding)
+			Expect(err).ToNot(HaveOccurred())
+
+			By("Login to the cluster by a user without a role")
+			cmd := exec.Command("bash", "-c", "oc whoami --show-server")
+			kubernetesApiEndpoint, err := cmd.CombinedOutput()
+			Expect(err).ToNot(HaveOccurred(), "Failed to get kubernetes api endpoint")
+
+			loginCommand := fmt.Sprintf("oc login -u %s -p %s %s", nonAdminUser, nonAdminUser, string(kubernetesApiEndpoint))
```
Comment (siserafin) [inline test/e2e/rbac/rbac_test.go]: login can be a helper function loginAsNonAdminUser(harness *e2e.Harness)

File: test/e2e/rbac/rbac_test.go [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+			cmd = exec.Command("bash", "-c", loginCommand)
+			err = cmd.Run()
+			Expect(err).ToNot(HaveOccurred())
+			defer ChangeContext("default")
+
+			method := login.LoginToAPIWithToken(harness)
+			Expect(method).ToNot(Equal(login.AuthDisabled))
+
+			By("Testing device operations - admin should have full access")
+			By("Testing creating a device")
```
Comment (siserafin) [inline test/e2e/rbac/rbac_test.go]: Resource creation and update steps are repeated . I think we could have a function like:
```
out, err := testResourceCreation(harness *e2e.Harness, resourceType, yamlFile string, flightCtlResources  []flightCtlResource)
Same for update.

File: test/e2e/rbac/rbac_test.go [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+			device.Metadata.Labels = &map[string]string{"test": "rbac-admin"}
+			deviceData, err := yaml.Marshal(&device)
+			Expect(err).ToNot(HaveOccurred())
+			logrus.Println("deviceData: ", string(deviceData))
+			out, err = harness.CLIWithStdin(string(deviceData), "apply", "-f", "-")
+			Expect(err).ToNot(HaveOccurred())
+			Expect(out).To(MatchRegexp(`200 OK`))
+
+			By("Testing getting a specific device")
+			_, err = harness.CLI("get", "device"+"/"+*deviceName)
```
Comment (siserafin) [inline test/e2e/rbac/rbac_test.go]: We have GetResourcesByName

File: test/e2e/rbac/rbac_test.go [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+				out, err = harness.CLI("apply", "-f", resourceYamls["device"])
+				return err
+			}).Should(BeNil(), "failed to create a Device")
+			Expect(out).To(MatchRegexp(`201 Created`))
+			device := harness.GetDeviceByYaml(resourceYamls["device"])
+			deviceName := device.Metadata.Name
+			flightCtlResources = append(flightCtlResources, flightCtlResource{ResourceType: "device", ResourceName: *deviceName})
+
+			By("Testing updating a device")
+			device.Metadata.Labels = &map[string]string{"test": "rbac-admin"}
```
Comment (siserafin) [inline test/e2e/rbac/rbac_test.go]: We could have a helper function:
func addTestLabel(resource interface{})

**Review by siserafin — COMMENTED** [review]
Looks good, I think we can have some optimizations, but the flow is ok.

File: test/e2e/rbac/rbac_test.go (hunk 7 above)
Comment (itzikb-redhat) [reply test/e2e/rbac/rbac_test.go]: Thanks! Using it for listing resources as well.

File: test/e2e/rbac/rbac_test.go [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+			Expect(err).ToNot(HaveOccurred(), "Admin should be able to delete role binding")
+
+			By("Testing listing repositories without a role - should not be able to list")
+			_, err = harness.GetResourcesByName("repositories")
+			Expect(err).To(HaveOccurred(), "A user without a role should not be able to list repositories")
+		})
+	})
+})
+
+func changeK8sContext(harness *e2e.Harness, k8sContext string) {
```
Comment (siserafin) [inline test/e2e/rbac/rbac_test.go]: Please move this function to harness, we might re-use it

File: test/e2e/rbac/rbac_test.go [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+	cmd := exec.Command("oc", "config", "use-context", k8sContext)
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		GinkgoWriter.Printf("❌ Failed to change context to %s: %v\n", k8sContext, err)
+	} else {
+		GinkgoWriter.Printf("✅ Changed context to %s: %s\n", k8sContext, output)
+	}
+}
+
+func deleteRoleBinding(ctx context.Context, client kubernetes.Interface, namespace string, roleBindingName string) error {
```
Comment (siserafin) [inline test/e2e/rbac/rbac_test.go]: All these role-related functions might be re-used. I think we could move these to a file test/harness/e2e/harness_rbac.go

File: test/e2e/rbac/rbac_test.go [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+		err := deleteRoleBinding(suiteCtx, harness.Cluster, flightCtlNs, roleBinding)
+		if err != nil {
+			logrus.Errorf("Failed to delete role binding %s: %v", roleBinding, err)
+		} else {
+			GinkgoWriter.Printf("Deleted role binding %s", roleBinding)
+		}
+	}
+}
+
+func createResource(harness *e2e.Harness, resourceType string, flightCtlResources []flightCtlResource) (string, []byte, error) {
```
Comment (siserafin) [inline test/e2e/rbac/rbac_test.go]: please move this function to harness

File: test/e2e/rbac/rbac_test.go [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+	changeK8sContext(harness, k8sContext)
+	cmd := exec.Command("bash", "-c", "oc whoami --show-server")
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return "", fmt.Errorf("Failed to get Kubernetes API endpoint: %v", err)
+	}
+	return strings.TrimSpace(string(output)), nil
+}
+
+func loginAsNonAdmin(harness *e2e.Harness, user string, password string, k8sContext string, k8sApiEndpoint string) error {
```
Comment (siserafin) [inline test/e2e/rbac/rbac_test.go]: this function can be moved to login.go

File: test/e2e/rbac/rbac_test.go [hunk 13]
```diff
@@ -0,0 +1,338 @@
...
+			return "", nil, err
+		}
+		return *resourceName, resourceData, nil
+	} else {
+		GinkgoWriter.Printf("Apply output: %s\n", applyOutput)
+		return "", nil, fmt.Errorf("Failed to create a %s", resourceType)
+	}
+}
+
+func getDefaultK8sContext() (string, error) {
```
Comment (siserafin) [inline test/e2e/rbac/rbac_test.go]: same comment here

File: test/e2e/rbac/rbac_test.go [hunk 14]
```diff
@@ -0,0 +1,338 @@
...
+	for _, context := range contexts {
+		if strings.Contains(context, "default") {
+			GinkgoWriter.Printf("🔍 [DEBUG] Found default context: %s\n", context)
+			return context, nil
+		}
+	}
+	return "", fmt.Errorf("no context with 'default' in name found")
+}
+
+func getK8sApiEndpoint(harness *e2e.Harness, k8sContext string) (string, error) {
```
Comment (siserafin) [inline test/e2e/rbac/rbac_test.go]: same comment here

File: test/e2e/rbac/rbac_suite_test.go [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+var (
+	flightCtlNs string
+)
+
+var _ = BeforeSuite(func() {
+	_, _, err := e2e.SetupWorkerHarness()
+	Expect(err).ToNot(HaveOccurred())
+
+	// Check if ACM is installed before running any tests
+	isAcmInstalled, err := util.IsAcmInstalled()
```
Comment (siserafin) [inline test/e2e/rbac/rbac_suite_test.go:27]: I think this test would work also in OCP + standalone. Maybe we can just check the context.

Comment (itzikb-redhat) [reply test/e2e/rbac/rbac_suite_test.go:27]: As it is right now we don't have user demouser1 in standalone

File: test/e2e/rbac/rbac_test.go (hunk 9 above)
Comment (itzikb-redhat) [reply test/e2e/rbac/rbac_test.go]: Done

File: test/e2e/rbac/rbac_test.go (hunk 10 above)
Comment (itzikb-redhat) [reply test/e2e/rbac/rbac_test.go]: Done

File: test/e2e/rbac/rbac_test.go (hunk 11 above)
Comment (itzikb-redhat) [reply test/e2e/rbac/rbac_test.go]: Done

File: test/e2e/rbac/rbac_test.go (hunk 14 above)
Comment (itzikb-redhat) [reply test/e2e/rbac/rbac_test.go]: Done

File: test/e2e/rbac/rbac_test.go (hunk 12 above)
Comment (itzikb-redhat) [reply test/e2e/rbac/rbac_test.go]: Done

**Comment by itzikb-redhat** [comment]
> So far it looks good! Is this tested in OCP?

It was tested locally with ACM.
One point:
A VM is created as part of the setup - NewTestHarnessWithoutVM is going to be removed and I haven't found another way

File: test/e2e/rbac/rbac_test.go [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+			Name:     adminRoleName,
+		}
+		adminRoleBinding.Subjects = []rbacv1.Subject{
+			{
+				Kind: "User",
+				Name: nonAdminUser,
+			},
+		}
+
+		It("should have full access to all resources and operations", Label("sanity", "83842"), func() {
```
Comment (rawagner) [inline test/e2e/rbac/rbac_test.go]: I'd make the test a bit more comprehensive by:
1. running the test scenario without role/rolebinding - and expect it to fail with 403
2. running the test again with role/rolebinding created (as you do now)
3. create the role/rolebinding in different NS and expect it to fail with 403
4. create the clusterrole/clusterolebinding and expect it to pass

**Review by rawagner — APPROVED** [review]

File: test/e2e/rbac/rbac_test.go [hunk 17]
```diff
@@ -0,0 +1,293 @@
...
+			Expect(err).ToNot(HaveOccurred(), "Admin should be able to delete cluster role")
+			err = harness.DeleteClusterRoleBinding(suiteCtx, harness.Cluster, createdAdminClusterRoleBinding.Name)
+			Expect(err).ToNot(HaveOccurred(), "Admin should be able to delete cluster role binding")
+		})
+	})
+})
+
+// testResourceOperations tests all CRUD operations for the given resource types
+// shouldSucceed determines whether the operations are expected to succeed or fail
+func testResourceOperations(ctx context.Context, harness *e2e.Harness, resourceTypes []string, shouldSucceed bool, testLabels *map[string]string, namespace string, operations []string) {
```
Comment (siserafin) [inline test/e2e/rbac/rbac_test.go]: Nice! I think this function should return output and error, and validate with Expect from the test.

Comment (siserafin) [reply test/e2e/rbac/rbac_test.go]: Maybe consider to move it to harness. I guess we can re-use it.

File: test/e2e/rbac/rbac_test.go [hunk 18]
```diff
@@ -0,0 +1,293 @@
...
+				}
+			}
+		}
+	}
+}
+
+// testReadOnlyResourceOperations tests read-only operations for the given resource types
+// for the given resource types
+// shouldSucceed determines whether the operations are expected to succeed or fail
+func testReadOnlyResourceOperations(harness *e2e.Harness, resourceTypes []string, shouldSucceed bool) {
```
Comment (siserafin) [inline test/e2e/rbac/rbac_test.go]: Same comment here....I would move the Expect to the test case.

File: test/harness/e2e/harness.go [hunk 19]
```diff
@@ -1928,3 +1964,107 @@ func (h *Harness) CreateGitRepositoryWithContent(repoName, filePath, content str
...
+	if err != nil {
+		GinkgoWriter.Printf("❌ Failed to change context to %s: %v\n", k8sContext, err)
+		return fmt.Errorf("failed to change context to %s: %w", k8sContext, err)
+	} else {
+		GinkgoWriter.Printf("✅ Changed context to %s: %s\n", k8sContext, output)
+		return nil
+	}
+}
+
+func (h *Harness) CreateResource(resourceType string) (string, string, []byte, error) {
```
Comment (siserafin) [inline test/harness/e2e/harness.go:1998]: Maybe I would call this function apply resource, since there are different way to create resources, here you are using the client

File: test/harness/e2e/harness.go [hunk 20]
```diff
@@ -1928,3 +1964,107 @@ func (h *Harness) CreateGitRepositoryWithContent(repoName, filePath, content str
...
+	}
+	return "", fmt.Errorf("no context with 'default' in name found")
+}
+
+// GetK8sApiEndpoint returns the API endpoint for a given K8s context
+func (h *Harness) GetK8sApiEndpoint(k8sContext string) (string, error) {
+	if err := h.ChangeK8sContext(k8sContext); err != nil {
+		return "", fmt.Errorf("failed to change K8s context to %s: %w", k8sContext, err)
+	}
+	cmd := exec.Command("bash", "-c", "oc whoami --show-server")
```
Comment (siserafin) [inline test/harness/e2e/harness.go:2084]: I would add a check that oc client is installed in this function.

File: test/e2e/rbac/rbac_test.go [hunk 21]
```diff
@@ -0,0 +1,293 @@
...
+			Name:     adminRoleName,
+		}
+		adminClusterRoleBinding.Subjects = []rbacv1.Subject{
+			{
+				Kind: "User",
+				Name: nonAdminUser,
+			},
+		}
+
+		It("should have access full access with an admin role", Label("sanity", "83842"), func() {
```
Comment (siserafin) [inline test/e2e/rbac/rbac_test.go]: Not sure If we should add the sanity label. This test is going to be skipped in the pre-PR checks, since ACM is not part of the env.

File: test/e2e/rbac/rbac_test.go (hunk 17 above)
Comment (itzikb-redhat) [reply test/e2e/rbac/rbac_test.go]: Moved

File: test/e2e/rbac/rbac_test.go (hunk 18 above)
Comment (itzikb-redhat) [reply test/e2e/rbac/rbac_test.go]: Moved

File: test/harness/e2e/harness.go (hunk 20 above)
Comment (itzikb-redhat) [reply test/harness/e2e/harness.go:2084]: Done
